dev-libs/msgpack: version bump to 1.4.2 and 2.2.1

MathyV opened the pull request:
An update to the latest version of 1.X and 2.X of the package. This was
already being worked on in PR #2974, I processed the comments and added
some extra things:

* Updated copyright
* EAPI version bump
* Updated license to Boost-1.0
* Added boost USE-flag
* Added examples USE-flag
* Added doc USE-flag
* Updated patches for cflags and static-libs
* Masked for testing

The new versions are masked for testing as the status of depending
packages is currently unknown. Bugs for the package maintainers will be
created.

Gentoo-Bug: https://bugs.gentoo.org/show_bug.cgi?id=601370

Package-Manager: Portage-2.3.3, Repoman-2.3.1

stefan-langenmaier commented on dev-libs/msgpack/msgpack-2.1.1.ebuild:
```diff
@@ -0,0 +1,83 @@
+		>=dev-cpp/gtest-1.6.0-r2[${MULTILIB_USEDEP}]
+		sys-libs/zlib[${MULTILIB_USEDEP}]
+	)
+	boost? ( dev-libs/boost[static-libs] )
```
I also did some more tests with the old pull request an I saw that in CMakeLists.txt it checks for the context component, so to build with boost support it needs the context use flag enabled.

SoapGentoo replied:
have you checked whether it links? I thought context is also a linking lib

MathyV replied:
Turns out you don't even need boost to build this, it is never even linked, all header-only, limited to msgpack's include directory. I'll remove the boost USE-flag as discussed with @SoapGentoo on IRC.

stefan-langenmaier commented:
Thanks for pushing this forward.

SoapGentoo commented on dev-libs/msgpack/msgpack-1.4.2.ebuild:
```diff
@@ -0,0 +1,83 @@
+	doc? ( app-doc/doxygen )
+"
+
+DOCS=( README.md )
```
do you need this or just want to inhibit installing other stuff implicitly?

MathyV replied:
Removed

SoapGentoo commented on dev-libs/msgpack/msgpack-1.4.2.ebuild:
```diff
@@ -0,0 +1,83 @@
+		if use doc ; then
+			local HTML_DOCS=( "${BUILD_DIR}"/docs/. )
+
+			mkdir docs
```
missing `|| die`

MathyV replied:
Fixed

SoapGentoo commented on dev-libs/msgpack/msgpack-1.4.2.ebuild:
```diff
@@ -0,0 +1,83 @@
+		fi
+
+		if use examples ; then
+			insinto /usr/share/doc/${PF}
```
`docinto examples`

MathyV replied:
Fixed

SoapGentoo commented on dev-libs/msgpack/msgpack-1.4.2.ebuild:
```diff
@@ -0,0 +1,83 @@
+
+			pushd "${WORKDIR}/${P}" > /dev/null
+			doins -r example
+			popd > /dev/null
```
scrap all of this and just do
`dodoc -r "${WORKDIR}/${P}/example/."`

MathyV replied:
Fixed

SoapGentoo commented on dev-libs/msgpack/msgpack-1.4.2.ebuild:
```diff
@@ -0,0 +1,83 @@
+			doins -r example
+			popd > /dev/null
+
+			docompress -x /usr/share/doc/${PF}/example
```
`docompress -x /usr/share/doc/${PF}/examples`

MathyV replied:
Fixed

MathyV commented on dev-libs/msgpack/msgpack-2.1.1.ebuild:
```diff
@@ -0,0 +1,83 @@
+		-DMSGPACK_BUILD_TESTS="$(usex test)"
+		-DMSGPACK_BOOST="$(usex boost)"
+		# Don't build the examples
+		-DMSGPACK_BUILD_EXAMPLES=OFF
```
I discussed this on IRC with the devs, they shouldn't be built (I actually did that first), just installed in the /usr/share/doc directory.

lluixhi commented:
Hmmm. I was actually looking into the same thing today (before seeing your pull request), and there are a couple of things that you missed -- first of all, I realized both msgpack-1.4.x and msgpack-2.x use the same SONAME, which is 2.0.0.

On the other hand, msgpack-2.x are the first versions to include the msgpack v2 c++ API in addition to the v1 api -- so I propose that we only add 2.1.1 and put it under the slotting "0/2". msgpack-1.4.2 appears to be redundant.

Currently, the cxx use flag does not do anything -- originally the C++ support built a separate library for C++, but now that library is header-only (as of 1.4.x) -- and the install target in CMakeLists.txt was not updated. Here's a patch that I wrote to fix that:
```
diff -Naur msgpack-2.1.1.orig/CMakeLists.txt msgpack-2.1.1/CMakeLists.txt
--- msgpack-2.1.1.orig/CMakeLists.txt	2017-01-09 04:36:00.000000000 -0800
+++ msgpack-2.1.1/CMakeLists.txt	2017-02-27 14:48:44.567218945 -0800
@@ -258,7 +258,10 @@
 INSTALL (TARGETS ${MSGPACK_INSTALLTARGETS} RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
                                            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
                                            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
-INSTALL (DIRECTORY include DESTINATION ${CMAKE_INSTALL_PREFIX})
+FOREACH (file ${msgpackc_HEADERS})
+    GET_FILENAME_COMPONENT (dir ${file} DIRECTORY)
+    INSTALL (FILES ${file} DESTINATION ${CMAKE_INSTALL_PREFIX}/${dir})
+ENDFOREACH ()
 IF (NOT MSVC)
     INSTALL (FILES ${CMAKE_CURRENT_BINARY_DIR}/msgpack.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
 ENDIF ()
```

In addition, the Boost support will only work if you use boost[static-libs], but that is fixable like so:
```
diff -Naur msgpack-2.1.1.orig/CMakeLists.txt msgpack-2.1.1/CMakeLists.txt
--- msgpack-2.1.1.orig/CMakeLists.txt	2017-02-27 14:57:53.946171756 -0800
+++ msgpack-2.1.1/CMakeLists.txt	2017-02-27 14:58:03.359257248 -0800
@@ -94,7 +94,6 @@
    SET (CMAKE_CXX_FLAGS "-DMSGPACK_USE_BOOST ${CMAKE_CXX_FLAGS}")
 ENDIF ()
 
-SET (Boost_USE_STATIC_LIBS        ON) # only find static libs
 SET (Boost_USE_MULTITHREADED      ON)
 SET (Boost_USE_STATIC_RUNTIME    OFF)
 FIND_PACKAGE (Boost COMPONENTS chrono context timer system)
```

Thirdly, I don't know if I agree with the way you're just bumping the `static` and `cflags` patches -- realistically we only need to be removing '-Werror -g -O3' from the CFLAGS, not deleting the entire string.
The static patch that was originally provided does work, though I have a different one that inverts the logic, which I think is closer to what we want..

MathyV commented:
@SoapGentoo I removed the $Id$

@lluixhi:
* I created 1.4.2 specifically because it is a specific version I required for usage with autobahn-cpp, a later version did not work so I guess there are still some big differences between 1.x and 2.x even though 1.4.2 indeed already seems to support the v2 API. The reason these updates are entering the tree masked is because they need testing. I tried to compile all the packages that depend on msgpack that are in the tree and they all compile fine but I'm in no position to test them, that is up to the package maintainers. If their tests show that it actually works we can avoid this slotting altogether since there won't be packages requiring it. I discussed this at length on IRC with the devs.
* can you create a pull request upstream to fix the MSGPACK_ENABLE_CXX option? I'd rather see an 'officially approved' patch for that.
* after some research it appears that boost isn't actually used to build the library itself, it is only ever used in the header-only part of the library, so the depending application needs to depend on the correct boost settings if they want to use that specific part of the library, not msgpack itself. After discussion on IRC I pulled the boost dependency. The only thing that does need it is like you said one of the examples, but since they aren't built and just documentation there is no issue as I expect a developer to figure this out on his own.
* I just updated the patch to work again on the latest version. If it is not correct then it is up to the package maintainers to comment and/or change it
* I updated the doxygen dependency

lluixhi commented:
Alright, thanks. I'll submit the `ENABLE_CXX` patch upstream.

lluixhi commented:
@MathyV
Patch was merged.
https://github.com/msgpack/msgpack-c/commit/791f55da35389059381043f09b5e02584da7a53b

stefan-langenmaier commented:
What is the strategy? Are we waiting till the patch is in a release?

mgorny commented:
Ping. Could you let us know whether this should be merged or are you intending to wait?

MathyV commented:
Since the patch was submitted upstream we could include it in commit since there is no new upstream release that includes it yet (but it works without too). I'll see if I have time to check this next week.

radhermit commented:
As noted previously, this still could use the addition of subslots due to the changing soname. I'm not following upstream or any projects that use msgpack atm so I'm unsure if we need to fully slot it yet.

Also, if/while it's masked someone could look into the revdeps and add slot ops on them so people under the default portage dep resolution scheme will get auto-rebuilds. Otherwise there will probably be a bug or two filed at some point after unmasking since people don't like using revdep-rebuild or similar anymore. :smile:

mgorny commented:
Ping again.
